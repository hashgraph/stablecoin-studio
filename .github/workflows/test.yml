name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test