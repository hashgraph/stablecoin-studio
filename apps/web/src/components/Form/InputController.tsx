import type {
	InputProps as ChakraInputProps,
	FormControlProps,
	FormLabelProps,
	StackProps,
} from '@chakra-ui/react';
import {
	Flex,
	FormControl,
	FormErrorMessage,
	Input,
	InputGroup,
	Link,
	Stack,
} from '@chakra-ui/react';
import type { Control, FieldValues, UseControllerProps } from 'react-hook-form';
import { Controller } from 'react-hook-form';
import type { ChangeEventHandler, ChangeEvent, FocusEvent, ReactNode } from 'react';
import InputLabel from './InputLabel';
import { InfoIcon } from '@chakra-ui/icons';

export interface InputControllerProps extends Omit<ChakraInputProps, 'name'> {
	rules?: UseControllerProps['rules'];
	control: Control<FieldValues>;
	label?: string;
	disabled?: boolean;
	onChangeAux?: ChangeEventHandler<HTMLInputElement>;
	onBlurAux?: (event: FocusEvent<HTMLInputElement>) => void;
	showErrors?: boolean;
	'data-testid'?: string;
	leftElement?: ReactNode;
	rightElement?: ReactNode;
	autoComplete?: 'on' | 'off';
	formStyle?: FormControlProps;
	labelStyle?: FormLabelProps;
	name: string;
	containerStyle?: StackProps;
	link?: string | null;
}

const InputController = ({
	name,
	rules,
	label,
	placeholder,
	control,
	isRequired = false,
	disabled = false,
	onChangeAux,
	onBlurAux,
	showErrors = true,
	'data-testid': dataTestId = name,
	leftElement,
	rightElement,
	formStyle,
	autoComplete = 'off',
	labelStyle,
	defaultValue,
	containerStyle,
	link,
	...props
}: InputControllerProps) => {
	return (
		<Controller
			control={control}
			name={name}
			defaultValue={defaultValue}
			rules={rules}
			render={({ field: { onChange, value = '' }, fieldState: { error } }) => {
				const onChangeCustom = (event: ChangeEvent<HTMLInputElement>) => {
					onChange(event);
					onChangeAux && onChangeAux(event);
				};

				const onBlurCustom = (event: FocusEvent<HTMLInputElement>) => {
					onBlurAux && onBlurAux(event);
				};

				return (
					<Stack w='full' {...containerStyle}>
						<FormControl isInvalid={!!error} {...formStyle}>
							<Flex justifyContent='space-between'>
								{label && (
									<InputLabel isRequired={isRequired} style={labelStyle}>
										{label}
									</InputLabel>
								)}
								{link && (
									<Link href={link} isExternal>
										<InfoIcon />
									</Link>
								)}
							</Flex>
							<InputGroup>
								{leftElement}
								<Input
									data-testid={dataTestId}
									name={name}
									placeholder={placeholder}
									value={value}
									disabled={disabled}
									onChange={onChangeCustom}
									onBlur={onBlurCustom}
									autoComplete={autoComplete}
									{...props}
								/>
								{rightElement}
							</InputGroup>
							{showErrors && error && (
								<FormErrorMessage data-testid='input-error-message'>
									{error.message}
								</FormErrorMessage>
							)}
						</FormControl>
					</Stack>
				);
			}}
		/>
	);
};

export default InputController;
