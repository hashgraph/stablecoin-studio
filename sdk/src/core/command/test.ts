/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
