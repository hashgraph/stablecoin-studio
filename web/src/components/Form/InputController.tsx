import type {
	InputProps as ChakraInputProps,
	FormControlProps,
	FormLabelProps,
} from '@chakra-ui/react';
import { FormControl, FormErrorMessage, Input, InputGroup, Stack } from '@chakra-ui/react';
import type { Control, FieldValues, UseControllerProps } from 'react-hook-form';
import { Controller } from 'react-hook-form';
import type { ChangeEventHandler, ChangeEvent, FocusEvent, ReactNode } from 'react';
import InputLabel from './InputLabel';

export interface InputControllerProps extends Omit<ChakraInputProps, 'name'> {
	rules?: UseControllerProps['rules'];
	control: Control<FieldValues>;
	label?: string;
	disabled?: boolean;
	onChangeAux?: ChangeEventHandler<HTMLInputElement>;
	onBlurAux?: (event: FocusEvent<HTMLInputElement>) => void;
	showErrors?: boolean;
	'data-testid'?: string;
	leftElement?: ReactNode;
	rightElement?: ReactNode;
	autoComplete?: 'on' | 'off';
	formStyle?: FormControlProps;
	labelStyle?: FormLabelProps;
	name: string;
}

const InputController = ({
	name,
	rules,
	label,
	placeholder,
	control,
	isRequired = false,
	disabled = false,
	onChangeAux,
	onBlurAux,
	showErrors = true,
	'data-testid': dataTestId = name,
	leftElement,
	rightElement,
	formStyle,
	autoComplete = 'off',
	labelStyle,
	...props
}: InputControllerProps) => {
	return (
		<Controller
			control={control}
			name={name}
			rules={rules}
			render={({ field: { onChange, value = '' }, fieldState: { error } }) => {
				const onChangeCustom = (event: ChangeEvent<HTMLInputElement>) => {
					onChange(event);
					onChangeAux && onChangeAux(event);
				};

				const onBlurCustom = (event: FocusEvent<HTMLInputElement>) => {
					onBlurAux && onBlurAux(event);
				};

				return (
					<Stack w='full'>
						<FormControl isInvalid={!!error} {...formStyle}>
							{label && (
								<InputLabel isRequired={isRequired} style={labelStyle}>
									{label}
								</InputLabel>
							)}
							<InputGroup>
								{leftElement}
								<Input
									data-testid={dataTestId}
									name={name}
									placeholder={placeholder}
									value={value}
									disabled={disabled}
									onChange={onChangeCustom}
									onBlur={onBlurCustom}
									autoComplete={autoComplete}
									{...props}
								/>
								{rightElement}
							</InputGroup>
							{showErrors && error && (
								<FormErrorMessage data-testid='input-error-message'>
									{error.message}
								</FormErrorMessage>
							)}
						</FormControl>
					</Stack>
				);
			}}
		/>
	);
};

export default InputController;
