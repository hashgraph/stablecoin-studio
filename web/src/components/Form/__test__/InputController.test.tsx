import type { UseFormReturn } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { act, fireEvent, waitFor } from '@testing-library/react';
import type { InputControllerProps } from '../InputController';
import InputController from '../InputController';
import { render } from '../../../test/index';

let form: UseFormReturn;

const RenderWithForm = (props: Omit<InputControllerProps, 'control'>) => {
	const localForm = useForm({
		mode: 'onChange',
		criteriaMode: 'all',
	});

	form = localForm;

	return <InputController control={localForm.control} {...props} />;
};

const defaultProps: Omit<InputControllerProps, 'control'> = {
	type: 'text',
	name: 'input-test',
	'data-testid': 'input',
	label: '',
};

const factoryComponent = (props: Partial<InputControllerProps> = defaultProps) => {
	return render(<RenderWithForm name='input-test' {...props} />);
};

describe(`<${InputController.name} />`, () => {
	describe('.render', () => {
		test('should render the component', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.asFragment()).toMatchSnapshot('WithLabel');
		});

		test('should show the label', () => {
			const component = factoryComponent({ label: 'test-label' });

			expect(component.getByText('test-label')).toBeInTheDocument();
		});

		test('should show the error message', async () => {
			const component = factoryComponent({ showErrors: true });

			const errorMessage = 'Expired';
			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.getByTestId('input-error-message')).toHaveTextContent(errorMessage);
			expect(component.asFragment()).toMatchSnapshot('WithError');
		});
		test('shouldnt show the error message if showErrors=false', async () => {
			const component = factoryComponent({ showErrors: false });

			const errorMessage = 'Expired';
			act(() => {
				form.setError(defaultProps.name, { type: 'expired', message: errorMessage });
			});

			expect(component.queryByTestId('input-error-message')).toBeNull();
		});
		test('should show placeholder', () => {
			const component = factoryComponent({
				placeholder: 'placeholder',
			});

			const input = component.getByTestId(defaultProps.name);
			expect(input.getAttribute('placeholder')).toBe('placeholder');
			expect(component.asFragment()).toMatchSnapshot('WithPlaceholder');
		});

		test('should be disabled if disabled=true', () => {
			const component = factoryComponent({
				disabled: true,
			});
			const input = component.getByTestId(defaultProps.name);

			expect(input).toBeDisabled();
			expect(component.asFragment()).toMatchSnapshot('Disabled');
		});

		test('should render the right element', () => {
			const component = factoryComponent({
				rightElement: <span data-testid='right-element' />,
			});
			const rightElement = component.getByTestId('right-element');

			expect(rightElement).toBeInTheDocument();
			expect(component.asFragment()).toMatchSnapshot('WithRightElement');
		});

		test('should render the left element', () => {
			const component = factoryComponent({
				leftElement: <span data-testid='left-element' />,
			});
			const leftElement = component.getByTestId('left-element');

			expect(leftElement).toBeInTheDocument();
			expect(component.asFragment()).toMatchSnapshot('WithLeftElement');
		});
	});
	describe('.events', () => {
		test('calls onBlurAux', async () => {
			const onBlurCustom = jest.fn();
			const component = factoryComponent({ onBlurAux: onBlurCustom });
			const input = component.getByTestId(defaultProps.name);

			fireEvent.blur(input);
			await waitFor(() => {
				expect(onBlurCustom).toHaveBeenCalledTimes(1);
			});
		});

		test('calls onChangeAux', async () => {
			const onChangeCustom = jest.fn();

			const component = factoryComponent({
				onChangeAux: onChangeCustom,
			});
			const input = component.getByTestId(defaultProps.name);

			fireEvent.change(input, { target: { value: 'test' } });

			await waitFor(() => {
				expect(onChangeCustom).toHaveBeenCalled();
			});
		});
	});
});
