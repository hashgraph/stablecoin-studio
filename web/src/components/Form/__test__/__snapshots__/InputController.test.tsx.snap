// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputController /> .render should be disabled if disabled=true: Disabled 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          disabled=""
          id="field-:r5:"
          name="input-test"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputController /> .render should render the component: WithLabel 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <label
        class="chakra-form__label css-1hfoxt1"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
          >
            test-label
          </p>
        </div>
      </label>
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          id="field-:r0:"
          name="input-test"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputController /> .render should render the left element: WithLeftElement 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <span
          data-testid="left-element"
        />
        <input
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          id="field-:r7:"
          name="input-test"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputController /> .render should render the right element: WithRightElement 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          id="field-:r6:"
          name="input-test"
          value=""
        />
        <span
          data-testid="right-element"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputController /> .render should show placeholder: WithPlaceholder 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          id="field-:r4:"
          name="input-test"
          placeholder="placeholder"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<InputController /> .render should show the error message: WithError 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          aria-describedby="field-:r2:-feedback"
          aria-invalid="true"
          autocomplete="off"
          class="chakra-input css-g75xpy"
          data-testid="input-test"
          id="field-:r2:"
          name="input-test"
          value=""
        />
      </div>
      <div
        aria-live="polite"
        class="chakra-form__error-message css-2e3jkk"
        data-testid="input-error-message"
        id="field-:r2:-feedback"
      >
        Expired
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
